pf/tests/schema_and_program_test.go

```diff
@@ -137,3 +144,151 @@ resources:
 	assert.Len(t, diffs, 1)
 	assert.Equal(t, "DIFF_SOME", diffs[0].Response.Changes.String())
 }
+
+func TestIDAttribute(t *testing.T) {
+	tests := []struct {
+		name                string
+		attribute           rschema.Attribute
+		computeIDField      string
+		schemaName          string
+		expectErrorContains string
+		expectedIDOutput    string
+	}{
+		{
+			name:             "Valid - Optional + Computed",
+			attribute:        rschema.StringAttribute{Optional: true, Computed: true},
+			computeIDField:   "otherId",
+			schemaName:       "otherId",
+			expectedIDOutput: "test-id",
+		},
+		{
+			name:             "Valid - Optional",
+			attribute:        rschema.StringAttribute{Optional: true},
+			computeIDField:   "otherId",
+			schemaName:       "otherId",
+			expectedIDOutput: "test-id",
+		},
+		{
+			name:             "Valid - Computed (with computeID & Name)",
+			attribute:        rschema.StringAttribute{Computed: true},
+			computeIDField:   "otherId",
+			schemaName:       "otherId",
+			expectedIDOutput: "test-id",
+		},
+		{
+			name:             "Valid - Computed (without computeID & Name)",
+			attribute:        rschema.StringAttribute{Computed: true},
+			expectedIDOutput: "test-id",
+		},
+		{
+			// ComputeID must point to an existing field
+			// or have StringAttribute.Name provided as well
+			name:                "Invalid - Computed error",
+			attribute:           rschema.StringAttribute{Computed: true},
+			computeIDField:      "otherId",
+			expectErrorContains: "Could not find required property 'otherId' in state",
+		},
+		{
+			// ComputeID with no "Name"
+			name:             "Valid - Computed id points to a different field",
+			attribute:        rschema.StringAttribute{Computed: true},
+			computeIDField:   "s",
+			expectedIDOutput: "hello",
+		},
+		// This one fails on checks during tfgen which we don't have a way of catching
+		// including the test here for completeness
```

Comment:
@VenelinMartinov 2c here on e2e testing features it would be nice to be able to assert on expected tfgen failures.

Reply:
Hm, we do run `generateSchema` as part of the test, so if that errors it should pop up in the test.

Which part is failing during tfgen for this case? Should the tests include that?

Reply:
The error is thrown, but not in a way that can be captured and asserted on in the test.

Reply:
Is it a panic? We could `recover` from it for now and if we find we need this often we can abstract it away in the test framework.

Reply:
The are a bunch of `require.NoError()`s through the chain of functions in the test framework. I couldn't find a way of capturing the error. I could turn the error into a panic instead, but I'm not sure it's worth it here.

Reply:
We could thread through an option parameter to `startPulumiProvider`, something like `expectedTfgenError` and pass it all the way to `genMetadata` and have that assert on the error. That'd keep the interface the same for all other users.

Reply:
Ok I think I figured out a way to assert on the error. Let me know what you think.